library/user.inc

```diff
@@ -9,139 +9,123 @@
-// $user is the user id number in the 'users' table
-// Returns true if setting exist and false if does not exist
+/**
+ * Check if effective user setting matches given value
+ * 
+ * @param string $label - Setting key
+ * @param string $value - Setting value
+ * @param int $user - user id number from users table
+ * @return boolean - true if setting exist and false if does not exist
+ */
 function checkUserSetting($label, $value, $user=NULL) {
 
-  // if no user id is sent, then use the currently logged in user
-  if (!isset($user)) {
-    $user = $_SESSION['authUserID'];
-  }
-
-  // Check for the user settings
-  //  If does not exist, then check for a default entry.
-  //  (return false if setting does not exist)
-  $result = sqlQuery("SELECT setting_value FROM user_settings " .
-    "WHERE setting_user=? AND setting_label=? AND setting_value=?", array($user, $label, $value) );
-  if (empty($result)) {
-    $result = sqlQuery("SELECT setting_value FROM user_settings " .
-      "WHERE setting_user='0' AND setting_label=? AND setting_value=?", array($label, $value) );
-    if (empty($result)) {
-      return false;
-    }
+  $curval = getUserSetting ($label, $user);
+  if (is_null($curval)) {
+    return false;
+  } else {
+    return ($curval === $value);
   }
-  
-  // Setting exist, so return true
-  return true;
 }
 
-//This will set a user setting
-// $name is the setting name in the 'users' table
-// $value is the setting value to be set in the 'users' table
-// $user is the user id number in the 'users' table
-// $createDefault is flag to create a default value if it does not yet exist
+/**
+ * Set a user setting 
+ * 
+ * @param string $label - Setting key
+ * @param string $value - Setting value
+ * @param int $user - user id number from users table
+ * @param boolean $createDefault - If no current global default value, create one. 
+ */
 function setUserSetting($label, $value, $user=NULL, $createDefault=TRUE) {
-  // if no user id is sent, then use the currently logged in user
-  if (!isset($user)) {
-    $user = $_SESSION['authUserID'];
-  }
+  
+  $user = effectiveUser($user);
 
-  $newCustom=false;
-  $newDefault=false;
+  $cur_value = getUserSetting($label, $user);
 
   // Check for a custom settings
-  $row = sqlQuery("SELECT setting_value FROM user_settings " .
-    "WHERE setting_user=? AND setting_label=?", array($user, $label) );
-  if (empty($row)) {
-    $newCustom=true;
-  }
-
-  // Set the custom setting
-  if ($newCustom) {
-    sqlStatement("INSERT INTO user_settings " .
-      "(setting_user, setting_label, setting_value) " .
+  if (is_null($cur_value)) {
+    sqlStatement("INSERT INTO user_settings(setting_user, setting_label, setting_value) " .
       "VALUES (?,?,?)", array($user, $label, $value) );
-  }
-  else {
+  } elseif ($cur_value !== $value) {
     sqlStatement("UPDATE user_settings SET setting_value=? " .
       "WHERE setting_user=? AND setting_label=?", array($value, $user, $label) );
   }
```

Comment:
Will above algorithm break if the getUserSetting is only returning the default value (ie. user 0)?